tfjs-layers/src/layers/nlp/tokenizers.ts

```diff
@@ -134,24 +136,338 @@ export abstract class Tokenizer extends Layer {
+
+    const pairRight = unfinishedWords.map(
+      (word, idx) => word[minPairRankIndices[idx] + 1]);
+
+    const mergedPairs = pairLeft.map((left, idx) => {
+      const right = pairRight[idx];
+      return [left, right].join('');
+    });
+    const unfinishedWordsIndices = mask
+      .map((_, idx) => idx)
+      .filter((_, idx) => mask[idx]);
+
+    const mergedPairIndices = unfinishedWordsIndices.map(
+      (index, idx) => [index, minPairRankIndices[idx]]);
+    const emptyStringIndices = unfinishedWordsIndices.map(
+      (index, idx) => [index, minPairRankIndices[idx] + 1]);
+
+    mergedPairIndices.forEach((indices, idx) => {
+      const [wordIdx, charIdx] = indices;
+      const mergedPair = mergedPairs[idx];
+      wordsStr[wordIdx][charIdx] = mergedPair;
+    });
+
+    emptyStringIndices.forEach((indices) => {
+      const [wordIdx, charIdx] = indices;
+      wordsStr[wordIdx][charIdx] = '';
+    });
+
+    words = wordsStr.map(word => tensor(word));
+    words = await removeStringsFromInputs(words, '');
+
+    return [words, mask];
+  }
+
+  private async bpeMerge(words: Tensor[]): Promise<Tensor[]> {
+    const numWords = words.length;
+
+    // Merge bytes.
+    function loopCondition(mask: boolean[]): boolean {
+      return mask.some(e => e);
+    }
+
+    const initialMask: boolean[] = Array(numWords).fill(true);
+
+    let mergedWords = words;
+    let mask = initialMask;
+    while (loopCondition(mask)) {
+      [mergedWords, mask] = await this.bpeMergeOneStep(mergedWords, mask);
+    }
+
+    return mergedWords;
+  }
+
+  /**
+   * Map token bytes to unicode using `byte2unicode`.
+   */
+  private async transformBytes(tokens: Tensor): Promise<Tensor[]> {
+    const tokensStr = await tokens.data() as unknown as string[];
+
+    const splitBytes = tokensStr.map(
+      token => tensor(token.split('').map(c => c.charCodeAt(0))));
+    const splitUnicode = await this.byte2Unicode.lookup(splitBytes);
+
+    return splitUnicode;
+  }
+
+  /**
+   * Process unseen tokens and add to cache.
+   */
+  private async bpeMergeAndUpdateCache(tokens: Tensor) {
+    const words = await this.transformBytes(tokens);
+    const tokenizedWordsTensor = await this.bpeMerge(words);
+    const tokenizedWords =
+      await tensorArrTo2DArr<string>(tokenizedWordsTensor);
+
+    // For each word, join all its token by a whitespace,
+    // e.g., ["dragon", "fly"] => "dragon fly" for hash purpose.
+    const joinedTokens = tokenizedWords.map(word => word.join(' '));
+
+    await this.cache.insert(tokens, joinedTokens);
+  }
+
+  tokenize(inputs: Tensor): Tensor[] {
+    this.bpeMergeAndUpdateCache;
+    const stringInputs = inputs.dataSync() as unknown as string[];
+    return stringInputs.map(input => tensor(input.split(' ')));
+  }
+
+  override detokenize(inputs: Tensor[]): Tensor {
+    const stringInputs = inputs.map(
+      input => input.dataSync() as unknown as string[]);
+    return tensor(stringInputs.map(str => str.join(' ')));
+  }
```

Comment:
I won't pay attention to these functions for this PR since you have another that's implementing them. However, it might be good to make them both throw an error just so people know they're not implemented yet.